lib/spack/spack/binary_distribution.py

```diff
@@ -1162,48 +1163,142 @@ def push(specs, push_url, specs_kwargs=None, **kwargs):
+           "specfile_path": "none-or-path-to-locally-saved-specfile",
+           "verify_requested": "true-if-not-unsigned",
+           "verfied": "true-if-binary-pkg-was-already-verified"
+       }
     """
     if not spack.mirror.MirrorCollection():
         tty.die("Please add a spack mirror to allow " +
                 "download of pre-compiled packages.")
 
     tarball = tarball_path_name(spec, '.spack')
+    specfile_prefix = tarball_name(spec, '.spec')
 
-    urls_to_try = []
+    mirrors_to_try = []
 
-    if preferred_mirrors:
-        for preferred_url in preferred_mirrors:
-            urls_to_try.append(url_util.join(
-                preferred_url, _build_cache_relative_path, tarball))
+    try_first = [i['mirror_url'] for i in mirrors_for_spec] if mirrors_for_spec else []
+    try_next = [
+        i.fetch_url for i in spack.mirror.MirrorCollection().values()
+        if i.fetch_url not in try_first
+    ]
 
-    for mirror in spack.mirror.MirrorCollection().values():
-        if not preferred_mirrors or mirror.fetch_url not in preferred_mirrors:
-            urls_to_try.append(url_util.join(
-                mirror.fetch_url, _build_cache_relative_path, tarball))
-
-    for try_url in urls_to_try:
-        # stage the tarball into standard place
-        stage = Stage(try_url, name="build_cache", keep=True)
-        stage.create()
-        try:
-            stage.fetch()
-            return stage.save_filename
-        except fs.FetchError:
-            continue
+    for url in try_first + try_next:
+        mirrors_to_try.append({
+            'specfile': url_util.join(url,
+                                      _build_cache_relative_path, specfile_prefix),
+            'spackfile': url_util.join(url,
+                                       _build_cache_relative_path, tarball)
+        })
+
+    tried_to_verify_sigs = []
+
+    # Assumes we care more about finding a spec file by preferred ext
+    # than by mirrory priority.  This can be made less complicated as
+    # we remove support for deprecated spec formats and buildcache layouts.
+    for ext in ['json.sig', 'json', 'yaml']:
+        for mirror_to_try in mirrors_to_try:
+            specfile_url = '{0}.{1}'.format(mirror_to_try['specfile'], ext)
+            spackfile_url = mirror_to_try['spackfile']
+            local_specfile_stage = try_fetch(specfile_url)
+            if local_specfile_stage:
+                local_specfile_path = local_specfile_stage.save_filename
+                signature_verified = False
+
+                if ext.endswith('.sig') and not unsigned:
+                    tried_to_verify_sigs.append(specfile_url)
+                    signature_verified = try_verify(local_specfile_path)
+                    if not signature_verified:
+                        tty.warn("Failed to verify: {0}".format(specfile_url))
+
+                if unsigned or signature_verified or not ext.endswith('.sig'):
```

Comment:
```suggestion
                if unsigned or signature_verified:
```

Please test this...

Reply:
@haampie If we have exhausted any/all clearsigned specs in the list of mirrors, we still could find that an unsigned spec file corresponds to an old-style tarball where the signature is an .asc file inside the outer tarball.   Including the extension check allows spack to continue to install from those old-style tarballs.

I've added a commit to fix the behavior you noticed though, so that spack does not install unsigned specs without the non-default `--no-check-signature`.   Please let me know what you think.

Reply:
Hm, I think error handling should happen earlier, cause now there are fatal errors even if another download from another mirror would work.

Reply:
I have seen the same. Even installing from sources is prevented when a signature could not be verified.

Reply:
Yes, this is true.  But to be fair, this PR didn't change the current behavior in that respect:

```
root@2a1be06efa0b:/work/old_spack# spack --version
0.18.0.dev0 (b9e3ee6dd07710f2a377e1624d1665da65ca834b)
root@2a1be06efa0b:/work/old_spack# spack install zlib
==> Installing zlib-1.2.12-c3adf66ccrqb3h4blnsfkgivzronprwm
==> No binary for zlib-1.2.12-c3adf66ccrqb3h4blnsfkgivzronprwm found: installing from source
==> Using cached archive: /work/old_spack/var/spack/cache/_source-cache/archive/91/91844808532e5ce316b3c010929493c0244f3d37593afd6de04f71821d5136d9.tar.gz
==> Applied patch /work/old_spack/var/spack/repos/builtin/packages/zlib/configure-cc.patch
==> zlib: Executing phase: 'install'
==> zlib: Successfully installed zlib-1.2.12-c3adf66ccrqb3h4blnsfkgivzronprwm
  Fetch: 0.01s.  Build: 1.77s.  Total: 1.78s.
[+] /work/old_spack/opt/spack/linux-ubuntu18.04-haswell/gcc-7.5.0/zlib-1.2.12-c3adf66ccrqb3h4blnsfkgivzronprwm
root@2a1be06efa0b:/work/old_spack# spack buildcache create -u -d /work/mirror /c3adf66cc
==> Pushing binary packages to file:///work/mirror/build_cache
root@2a1be06efa0b:/work/old_spack# spack uninstall -y /c3adf66ccrqb
==> Successfully uninstalled zlib@1.2.12%gcc@7.5.0+optimize+pic+shared patches=0d38234 arch=linux-ubuntu18.04-haswell/c3adf66
root@2a1be06efa0b:/work/old_spack# spack mirror add binaries file:///work/mirror
root@2a1be06efa0b:/work/old_spack# spack install zlib
==> Installing zlib-1.2.12-c3adf66ccrqb3h4blnsfkgivzronprwm
==> Extracting zlib-1.2.12-c3adf66ccrqb3h4blnsfkgivzronprwm from binary cache
==> Error: Failed to install zlib due to NoVerifyException: Package spec file failed signature verification.
Use spack buildcache keys to download and install a key for verification from the mirror.
root@2a1be06efa0b:/work/old_spack#
```

The suggestion there is wrong because the package is unsigned to begin with.  It also doesn't decide to try other mirrors that could succeed, or install from source either.

But you're right that my PR doesn't improve on that at all, and buildcache could use a lot of improvement.

Reply:
Yeah, sorry @scottwittenburg This may end up collecting a lot of comments which are really technical debts in buildcaches, but ought to be fixed before we advise using them by default.

Reply:
Yes, agreed.